tests/python_package_test/test_basic.py

```diff
@@ -510,19 +510,23 @@ def test_choose_param_value():
     assert original_params == expected_params
 
 
-@pytest.mark.skipif(not PANDAS_INSTALLED, reason='pandas is not installed')
-@pytest.mark.parametrize(
-    'y',
-    [
-        np.random.rand(10),
-        np.random.rand(10, 1),
-        pd_Series(np.random.rand(10)),
-        pd_Series(['a', 'b']),
-        [1] * 10,
-        [[1], [2]]
-    ])
+@pytest.mark.parametrize('collection', ['1d_np', '2d_np', 'pd_float', 'pd_str', '1d_list', '2d_list'])
```

Comment:
I fear this construct might be hard to understand if the reason behind it is unknown. 
Would it be an option to not change the test at all, but adapt the dummy implementation for pandas in lightgbm/compat.py

Current:
```python
try:
    from pandas import DataFrame as pd_DataFrame
    from pandas import Series as pd_Series
    from pandas import concat
    from pandas.api.types import is_sparse as is_dtype_sparse
    PANDAS_INSTALLED = True
except ImportError:
    PANDAS_INSTALLED = False

    class pd_Series:  # type: ignore
        """Dummy class for pandas.Series."""

        pass

    class pd_DataFrame:  # type: ignore
        """Dummy class for pandas.DataFrame."""

        pass
```

Proposal:
```python
try:
    from pandas import DataFrame as pd_DataFrame
    from pandas import Series as pd_Series
    from pandas import concat
    from pandas.api.types import is_sparse as is_dtype_sparse
    PANDAS_INSTALLED = True
except ImportError:
    PANDAS_INSTALLED = False

    class pd_Series:  # type: ignore
        """Dummy class for pandas.Series."""
        def __init__(self, *args):
            pass

    class pd_DataFrame:  # type: ignore
        """Dummy class for pandas.DataFrame."""
        def __init__(self, *args):
            pass
```

Reply:
I agree with @strobelTha . I believe that making all our dummy classes from `compat.py` creatable with any combination of arguments by adding catch-all constructor will be useful
```python
class pd_Series:  # type: ignore
    """Dummy class for pandas.Series."""
    def __init__(self, *args, **kwargs):
        pass
```

Reply:
I like the refactor because of the following reasons:
* Moving the collection definition inside the function instead of having that big parametrize seems cleaner.
* Now the test will run for lists and numpy arrays when pandas isn't installed, whereas previously the whole test was (supposed to be) skipped.
* In case the test fails the message is more descriptive, i.e. `FAILED tests/python_package_test/test_basic.py::test_list_to_1d_numpy[float32-2d_np]` vs `FAILED tests/python_package_test/test_basic.py::test_list_to_1d_numpy[float64-y1]`.

> ```
> class pd_Series:  # type: ignore
>     """Dummy class for pandas.Series."""
>     def __init__(self, *args, **kwargs):
>         pass
> ```

@strobelTha would you like to contribute those changes?

Reply:
@jmoralez I liked reasons `#2` and `#3` you listed above. Thanks a lot for emphasizing advantages of the current refactor.

And indeed, we can add constructors for our dummy classes in a separate PR.